File: flutter_obdii/test/viewmodels/gauges_viewmodel_test.dart
// Port of GaugesViewModelTests.swift — Jim Mittler
// Tests tile generation from enabled gauges, tile identity,
// stats association, and unit change rebuild.

import 'dart:async';

import 'package:flutter_test/flutter_test.dart';
import 'package:flutter_obd2/flutter_obd2.dart' as obd2lib;

import 'package:flutter_obdii/core/config_data.dart';
import 'package:flutter_obdii/core/obd_connection_manager.dart';
import 'package:flutter_obdii/core/pid_interest_registry.dart';
import 'package:flutter_obdii/core/pid_store.dart';
import 'package:flutter_obdii/models/obdii_pid.dart';
import 'package:flutter_obdii/viewmodels/gauges_viewmodel.dart';

// ─────────────────────────────────────────────
// Mocks (1:1 with Swift mocks)
// ─────────────────────────────────────────────

class MockPIDProvider implements PidListProviding {
  List<ObdiiPid> _pids = [];
  @override
  List<ObdiiPid> get pids => _pids;

  final _ctrl = StreamController<List<ObdiiPid>>.broadcast();

  @override
  Stream<List<ObdiiPid>> get pidsStream => _ctrl.stream;

  void send(List<ObdiiPid> pids) {
    _pids = pids;
    _ctrl.add(pids);
  }

  void dispose() => _ctrl.close();
}

class MockStatsProvider implements PidStatsProviding {
  @override
  Map<String, PIDStats> pidStats = {};

  final _ctrl = StreamController<Map<String, PIDStats>>.broadcast();

  @override
  Stream<Map<String, PIDStats>> get pidStatsStream => _ctrl.stream;

  @override
  PIDStats? statsFor(String cmd) => pidStats[cmd];

  void send(Map<String, PIDStats> stats) {
    pidStats = stats;
    _ctrl.add(stats);
  }

  void dispose() => _ctrl.close();
}

class MockUnitsProvider implements UnitsProviding {
  MeasurementUnit _units = MeasurementUnit.metric;

  @override
  MeasurementUnit get units => _units;

  final _ctrl = StreamController<MeasurementUnit>.broadcast();

  @override
  Stream<MeasurementUnit> get unitsStream => _ctrl.stream;

  void send(MeasurementUnit u) {
    _units = u;
    _ctrl.add(u);
  }

  void dispose() => _ctrl.close();
}

// Helper to build a gauge PID
ObdiiPid makeGaugePid({
  String id = 'pid_rpm',
  bool enabled = true,
  String label = 'RPM',
  String name = 'Engine RPM',
  String pidCommand = '010C',
  String units = 'RPM',
}) {
  return ObdiiPid(
    id: id,
    enabled: enabled,
    label: label,
    name: name,
    pidCommand: pidCommand,
    units: units,
    kind: ObdPidKind.gauge,
  );
}

void main() {
  late MockPIDProvider pidProvider;
  late MockStatsProvider statsProvider;
  late MockUnitsProvider unitsProvider;
  late PidInterestRegistry interestRegistry;
  late GaugesViewModel viewModel;

  setUp(() {
    pidProvider = MockPIDProvider();
    statsProvider = MockStatsProvider();
    unitsProvider = MockUnitsProvider();
    interestRegistry = PidInterestRegistry();
    viewModel = GaugesViewModel(
      pidProvider: pidProvider,
      statsProvider: statsProvider,
      unitsProvider: unitsProvider,
      interestRegistry: interestRegistry,
    );
  });

  tearDown(() {
    viewModel.dispose();
    pidProvider.dispose();
    statsProvider.dispose();
    unitsProvider.dispose();
  });

  // ── Initialization ──────────────────────────────

  test('testInitializationNotNullStartsEmpty', () {
    expect(viewModel, isNotNull);
    expect(viewModel.tiles, isEmpty);
  });

  // ── Tiles match enabled gauges ──────────────────

  test('testTilesIncludeOnlyEnabledGauges', () async {
    final pidEnabled = makeGaugePid(id: 'rpm', pidCommand: '010C');
    final pidDisabled =
        makeGaugePid(id: 'spd', enabled: false, label: 'Speed', pidCommand: '010D');
    pidProvider.send([pidEnabled, pidDisabled]);

    await Future.delayed(const Duration(milliseconds: 100));

    expect(viewModel.tiles.length, 1);
    expect(viewModel.tiles.first.pid.id, 'rpm');
  });

  test('testAllTilesHaveEnabledGaugePIDs', () async {
    final pid1 = makeGaugePid(id: 'a', pidCommand: '010C');
    final pid2 = makeGaugePid(id: 'b', pidCommand: '010D');
    pidProvider.send([pid1, pid2]);

    await Future.delayed(const Duration(milliseconds: 100));

    for (final tile in viewModel.tiles) {
      expect(tile.pid.enabled, isTrue);
      expect(tile.pid.kind, ObdPidKind.gauge);
    }
  });

  // ── Tile identity ───────────────────────────────

  test('testAllTileIDsAreNonEmpty', () async {
    pidProvider.send([makeGaugePid()]);
    await Future.delayed(const Duration(milliseconds: 100));

    for (final tile in viewModel.tiles) {
      expect(tile.id.isNotEmpty, isTrue);
    }
  });

  test('testAllTileIDsAreUnique', () async {
    pidProvider.send([
      makeGaugePid(id: 'x', pidCommand: '010C'),
      makeGaugePid(id: 'y', pidCommand: '010D'),
      makeGaugePid(id: 'z', pidCommand: '0105'),
    ]);
    await Future.delayed(const Duration(milliseconds: 100));

    final ids = viewModel.tiles.map((t) => t.id).toSet();
    expect(ids.length, viewModel.tiles.length);
  });

  // ── Stats association ───────────────────────────

  test('testTilesHaveMatchingStatsWhenProviderEmitsThem', () async {
    pidProvider.send([makeGaugePid(id: 'rpm', pidCommand: '010C')]);

    final result = obd2lib.MeasurementResult(1500.0, obd2lib.Unit.rpm);
    final stats = PIDStats(pid: '010C', latest: result);
    statsProvider.send({'010C': stats});

    await Future.delayed(const Duration(milliseconds: 100));

    expect(viewModel.tiles.length, 1);
    expect(viewModel.tiles.first.stats?.latest.value, closeTo(1500.0, 0.001));
  });

  // ── Unit change ─────────────────────────────────

  test('testUnitChangeRebuildsTilesWithSameIDs', () async {
    final pid = makeGaugePid(id: 'spd', pidCommand: '010D', units: 'km/h');
    pidProvider.send([pid]);

    final result = obd2lib.MeasurementResult(100.0, obd2lib.Unit.kilometersPerHour);
    statsProvider.send({'010D': PIDStats(pid: '010D', latest: result)});

    await Future.delayed(const Duration(milliseconds: 100));
    final firstIds = viewModel.tiles.map((t) => t.id).toList();

    unitsProvider.send(MeasurementUnit.imperial);
    await Future.delayed(const Duration(milliseconds: 100));

    final secondIds = viewModel.tiles.map((t) => t.id).toList();
    expect(firstIds, equals(secondIds));
    expect(viewModel.tiles.first.stats, isNotNull);
  });

  // ── Visibility-driven interest registration ─────

  test('testSetvisibleTrueRegistersEnabledGaugePIDInterests', () async {
    pidProvider.send([
      makeGaugePid(id: 'rpm', pidCommand: '010C'),
      makeGaugePid(id: 'spd', pidCommand: '010D'),
    ]);
    await Future.delayed(const Duration(milliseconds: 50));

    viewModel.setVisible(true);
    await Future.delayed(const Duration(milliseconds: 50));

    expect(interestRegistry.interested, containsAll(<String>{'010C', '010D'}));
  });

  test('testSetvisibleFalseClearsPreviouslyRegisteredInterests', () async {
    pidProvider.send([makeGaugePid(id: 'rpm', pidCommand: '010C')]);
    await Future.delayed(const Duration(milliseconds: 50));
    viewModel.setVisible(true);
    await Future.delayed(const Duration(milliseconds: 50));
    expect(interestRegistry.interested.contains('010C'), isTrue);

    viewModel.setVisible(false);
    await Future.delayed(const Duration(milliseconds: 50));
    expect(interestRegistry.interested.contains('010C'), isFalse);
  });

  test('testHiddenViewDoesNotRetainInterestsAfterPIDListChanges', () async {
    viewModel.setVisible(false);
    pidProvider.send([makeGaugePid(id: 'rpm', pidCommand: '010C')]);
    await Future.delayed(const Duration(milliseconds: 50));
    expect(interestRegistry.interested, isEmpty);
  });

  test('testSetvisibleTrueWithNoTilesKeepsInterestEmpty', () async {
    viewModel.setVisible(true);
    await Future.delayed(const Duration(milliseconds: 50));
    expect(interestRegistry.interested, isEmpty);
  });

  test('testStatsUpdateForUnknownPIDDoesNotCreateExtraTile', () async {
    pidProvider.send([makeGaugePid(id: 'rpm', pidCommand: '010C')]);
    statsProvider.send({
      '010D': PIDStats(
        pid: '010D',
        latest: obd2lib.MeasurementResult(40, obd2lib.Unit.kilometersPerHour),
      ),
    });
    await Future.delayed(const Duration(milliseconds: 100));
    expect(viewModel.tiles.length, 1);
    expect(viewModel.tiles.first.pid.pidCommand, '010C');
  });
}

File: kotlin_obdii/app/src/test/java/com/rheosoft/obdii/viewmodels/GaugesViewModelTest.kt
package com.rheosoft.obdii.viewmodels

import com.rheosoft.obdii.core.DefaultPidStore
import com.rheosoft.obdii.core.GaugesDisplayMode
import com.rheosoft.obdii.core.MeasurementResult
import com.rheosoft.obdii.core.PIDStats
import com.rheosoft.obdii.core.PidInterestRegistry
import com.rheosoft.obdii.core.PidStatsProviding
import com.rheosoft.obdii.core.UnitsProviding
import com.rheosoft.obdii.core.MeasurementUnit
import com.rheosoft.obdii.models.ObdPidKind
import com.rheosoft.obdii.models.ObdiiPid
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.test.runTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

private class MockStats : PidStatsProviding {
    private val flow = MutableStateFlow<Map<String, PIDStats>>(emptyMap())
    override val pidStats: Map<String, PIDStats> get() = flow.value
    override fun statsFor(pidCommand: String): PIDStats? = flow.value[pidCommand]
    override val pidStatsStream: StateFlow<Map<String, PIDStats>> = flow
    fun send(value: Map<String, PIDStats>) { flow.value = value }
}

private class MockUnits : UnitsProviding {
    private val flow = MutableStateFlow(MeasurementUnit.Metric)
    private val modeFlow = MutableStateFlow(GaugesDisplayMode.gauges)
    override val units: MeasurementUnit get() = flow.value
    override var gaugesDisplayMode: GaugesDisplayMode
        get() = modeFlow.value
        set(value) { modeFlow.value = value }
    override val unitsStream: StateFlow<MeasurementUnit> = flow
    override val gaugesDisplayModeStream: StateFlow<GaugesDisplayMode> = modeFlow
}

class GaugesViewModelTest {
    @Test
    fun `tiles include only enabled gauges`() = runTest {
        DefaultPidStore.resetForTests()
        DefaultPidStore.seededPidsProvider = {
            listOf(
                ObdiiPid("rpm", true, "RPM", "Engine RPM", "010C", units = "RPM", kind = ObdPidKind.gauge),
                ObdiiPid("spd", false, "Speed", "Vehicle Speed", "010D", units = "km/h", kind = ObdPidKind.gauge),
            )
        }
        DefaultPidStore.load()
        val vm = GaugesViewModel(DefaultPidStore, MockStats(), MockUnits(), PidInterestRegistry())
        Thread.sleep(50)
        assertEquals(1, vm.tiles.size)
        assertEquals("rpm", vm.tiles.first().pid.id)
    }

    @Test
    fun `visible registers interests`() = runTest {
        DefaultPidStore.resetForTests()
        DefaultPidStore.seededPidsProvider = {
            listOf(ObdiiPid("rpm", true, "RPM", "Engine RPM", "010C", units = "RPM", kind = ObdPidKind.gauge))
        }
        DefaultPidStore.load()
        val registry = PidInterestRegistry()
        val vm = GaugesViewModel(DefaultPidStore, MockStats(), MockUnits(), registry)
        Thread.sleep(50)
        vm.setVisible(true)
        Thread.sleep(50)
        assertTrue(registry.interested.contains("010C"))

        vm.setVisible(false)
        Thread.sleep(50)
        assertTrue(registry.interested.isEmpty())
    }

    @Test
    fun `setDisplayMode updates provider`() {
        val units = MockUnits()
        val vm = GaugesViewModel(DefaultPidStore, MockStats(), units, PidInterestRegistry())
        vm.setDisplayMode(GaugesDisplayMode.list)
        assertEquals(GaugesDisplayMode.list, units.gaugesDisplayMode)
    }

    @Test
    fun `isEmpty reflect tiles`() = runTest {
        DefaultPidStore.resetForTests()
        DefaultPidStore.seededPidsProvider = { emptyList() }
        DefaultPidStore.load()
        val vm = GaugesViewModel(DefaultPidStore, MockStats(), MockUnits(), PidInterestRegistry())
        assertTrue(vm.isEmpty)
    }

    @Test
    fun `moveEnabled calls store`() = runTest {
        val vm = GaugesViewModel(DefaultPidStore, MockStats(), MockUnits(), PidInterestRegistry())
        // DefaultPidStore is a PidStore
        vm.moveEnabled(0, 1)
        // Just verify it doesn't crash on JVM
    }

    @Test
    fun `moveEnabled handles non-PidStore gracefully`() = runTest {
        val mockProvider = object : com.rheosoft.obdii.core.PidListProviding {
            override val pids: List<ObdiiPid> = emptyList()
            override val pidsStream: StateFlow<List<ObdiiPid>> = MutableStateFlow(emptyList())
        }
        val vm = GaugesViewModel(mockProvider, MockStats(), MockUnits(), PidInterestRegistry())
        vm.moveEnabled(0, 1) // Should return safely
    }
}
